backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java

```diff
@@ -0,0 +1,229 @@
+			builder.locationId(bpLocationId.getRepoId());
+		}
+		final ExternalId externalId = relation.getExternalId();
+		if (externalId != null)
+		{
+			builder.externalId(JsonExternalId.ofOrNull(externalId.getValue()));
+		}
+
+		return builder.build();
+	}
+
+	public void createOrUpdateRelations(@NonNull final OrgId orgId,
+			@NonNull final IdentifierString bpartnerIdentifier,
+			@Nullable final IdentifierString locationIdentifier,
+			@NonNull final List<JsonRequestBPRelationTarget> relatesTo)
+	{
+		final Optional<BPartnerComposite> bPartnerComposite = jsonServiceFactory.createRetriever().getBPartnerComposite(orgId, bpartnerIdentifier);
+		final BPartnerComposite bpartner = Check.assumePresent(bPartnerComposite, "No bpartner found for identifier: " + bpartnerIdentifier);
+
+		final Optional<BPartnerLocation> location = getBpartnerLocation(bpartnerIdentifier, locationIdentifier, bpartner);
+		final Stream<BPRelation> relations = relatesTo.stream().map(relatedBp -> fromJson(bpartner, location, relatedBp));
+
+		relations.forEach(relation -> bpRelationDAO.saveOrUpdate(orgId, relation));
+	}
+
+	@NonNull
+	private Optional<BPartnerLocation> getBpartnerLocation(final @NonNull IdentifierString bpartnerIdentifier, final @Nullable IdentifierString locationIdentifier, @NonNull final BPartnerComposite bPartnerComposite)
+	{
+		final IdentifierString lookupLocationIdentifier = (locationIdentifier == null && IdentifierString.Type.GLN.equals(bpartnerIdentifier.getType()))
+				? bpartnerIdentifier
+				: locationIdentifier;
+		return bPartnerComposite.extractLocation(bPartnerLocation -> locationMatchesIdentifier(bPartnerLocation, lookupLocationIdentifier));
+	}
+
+	private boolean locationMatchesIdentifier(@NonNull final BPartnerLocation bPartnerLocation, @Nullable final IdentifierString locationIdentifier)
+	{
+		if (locationIdentifier == null)
+		{
+			return false;
+		}
+		switch (locationIdentifier.getType())
+		{
+			case EXTERNAL_ID:
+				return Objects.equals(bPartnerLocation.getExternalId(), locationIdentifier.asExternalId());
+			case GLN:
+				return GLN.equals(bPartnerLocation.getGln(), locationIdentifier.asGLN());
+			case METASFRESH_ID:
+				return MetasfreshId.equals(locationIdentifier.asMetasfreshId(), MetasfreshId.of(bPartnerLocation.getId()));
+			default:
+				// note: currently, "val-" is not supported with bpartner locations
+				throw new AdempiereException("Unexpected type=" + locationIdentifier.getType());
+		}
+	}
+
+	private BPRelation fromJson(final BPartnerComposite sourceBPartner, final Optional<BPartnerLocation> location, final JsonRequestBPRelationTarget relatedBp)
+	{
+		final BPartnerId bpartnerId = sourceBPartner.getBpartner().getId();
+		final IdentifierString targetBpIdentifier = IdentifierString.of(relatedBp.getTargetBpartnerIdentifier());
+		final BPartnerComposite targetBpartnerComposite =
+				Check.assumePresent(jsonServiceFactory.createRetriever().getBPartnerComposite(sourceBPartner.getOrgId(), targetBpIdentifier)
```

Comment:
please throw a MissingResourceException instead

Reply:
Changed to throw a MissingResourceException.